config/src/main/java/org/springframework/security/config/annotation/web/configurers/PermitAllSupport.java

```diff
@@ -48,11 +48,23 @@ static void permitAll(HttpSecurityBuilder<? extends HttpSecurityBuilder<?>> http
 			RequestMatcher... requestMatchers) {
 		ExpressionUrlAuthorizationConfigurer<?> configurer = http
 				.getConfigurer(ExpressionUrlAuthorizationConfigurer.class);
-		Assert.state(configurer != null, "permitAll only works with HttpSecurity.authorizeRequests()");
+		AuthorizeHttpRequestsConfigurer<?> httpConfigurer = http.getConfigurer(AuthorizeHttpRequestsConfigurer.class);
+
+		boolean oneConfigurerPresent = (configurer != null && httpConfigurer == null)
+				|| (configurer == null && httpConfigurer != null);
+		Assert.state(oneConfigurerPresent,
+				"permitAll only works with HttpSecurity.authorizeRequests() or HttpSecurity.authorizeHttpRequests()");
```

Comment:
I think this message could be a little clearer. Maybe:

```suggestion
				"permitAll only works with either HttpSecurity.authorizeRequests() or HttpSecurity.authorizeHttpRequests(). " +
				"Please define one or the other but not both.");
```